oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/Optimizer.kt

```diff
@@ -0,0 +1,358 @@
+            }
+        }
+    }
+
+    fun cancelAndStop() {
+        GlobalScope.launch {
+            stopOptimization()
+            for ((name, sim) in simulationsByName) {
+                val message = OASISQueryDTO.newBuilder().setCancelRequest(
+                        OASISQueryDTO.SimulationCancelRequest.newBuilder().setName(name)
+                ).build()
+                sim.input.onNext(message)
+            }
+        }
+    }
+
+    private suspend fun syncConfigFor(simName: String) {
+        val sim = simulationsByName.getValue(simName)
+        val message = OASISQueryDTO.newBuilder().setNodeStatusRequest(
+                OASISQueryDTO.NodeStatusUpdateRequest.newBuilder().setName(simName)
+        ).build()
+
+        sim.input.onNext(message)
+
+        val result = select<Either<Simulation, Message>> {
+            sim.update.onReceive { it -> Either.Left(updateFromResponse(it)) }
+            sim.error.onReceive { it ->
+                Either.Right(
+                        Message(LocalDateTime.now(), it.name, "Error update node $simName due to ${it.message} :\n${it.exception}")
+                )
+            }
+        }
+
+        if (result.isLeft()) {
+            simulationsByName += simName to result.left().get()
+        } else {
+            viewData.addMessage(result.right().get())
+        }
+    }
+
+
+    override fun stopOptimization(request: StopOptimizationCommandDTO, responseObserver: StreamObserver<StopOptimizationResponseDTO>) = responseObserver.consume {
+        stopOptimization()
+        StopOptimizationResponseDTO.newBuilder().setMessage("Stopped").setStopped(true).build()
+    }
+
+
+    override fun offerSimulationResult(request: SimulationResponseDTO, responseObserver: StreamObserver<SimulationResultConfirmDTO>) = responseObserver.consume {
+        val output = simulationsByName[request.name]?.output
+        if (output != null) {
+            output.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
+        SimulationResultConfirmDTO.newBuilder().build()
+    }
+
+    override fun offerErrorResult(request: ErrorResponseDTO, responseObserver: StreamObserver<ErrorConfirmDTO>) = responseObserver.consume {
+        val error = simulationsByName[request.name]?.error
+        if (error != null) {
+            error.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
+        ErrorConfirmDTO.newBuilder().build()
+    }
+
+
+    override fun offerSimulationConfig(request: NodeStatusCommandOrResponseDTO, responseObserver: StreamObserver<NodeChangeConfirmDTO>) = responseObserver.consume {
+        val update = simulationsByName[request.name]?.update
+
+        if (update != null) {
+            update.offer(request)
+        } else {
+            throw IllegalStateException("no simulation '${request.name}' or buffer full")
+        }
```

Comment:
I dont like this pattern we've got:

0. it seems like a lot of copy-paste
2. do we know what happens to this illegal state exception? What does GRPC do with it?
1. if `offer` returns false then we silently simply fail to send a reponse
2. `offer` applies no back-pressure: if the `update` channel is clogged up then you could easily have this system drop messages when the load gets high.

I think we should work together on deciding what to do with this kind of message marshalling.

Reply:
Sure, I wasn't put any thoughts on these. Just couple note not to answering your question above one by one. 
1. if channel is null, at this moment it means it was unregistered. 
2. unregistered != unavailable, so this will only cover the case where we somehow unregister the node during an evaluation. If the program dead or freezed, this will still try to do the offer.
3. For the exception, GRPC will throw the execution on the client side as well. 
4. For the case if we are not able to capture the result. Since the exception will follow on the client side, it will have its own respond. Currently, it will try to offer the error due to the exception thrown. Should we thinking about retry kind of flow, then we need to thinking the step to let user  recover a plugin in case of a freeze or a restart .

Reply:
> it seems liine a lot of copy-paste

any progress here?

---

>> * do we know what happens to this illegal state exception? What does GRPC do with it?
>* For the exception, GRPC will throw the execution on the client side as well.
>* For the case if we are not able to capture the result. Since the exception will follow on the client side, it will have its own respond. Currently, it will try to offer the error due to the exception thrown. Should we thinking about retry kind of flow, then we need to thinking the step to let user  recover a plugin in case of a freeze or a restart .

In other words, because we're in a `consume` block, the exception will hit an `onError`, and appear as a thrown exception to the caller of the API function on the client side. This will likely trigger the clients own default exception handler, likely leading to an `offerErrorResult` API call.

Makes sense to me. Can we have an interceptor log this with a nice message? If we had an outbound listener bouncing to j.u.logging, I would expect to see something like

```
[2019-02-25 11:05 AM] com.empowerops.volition.Optimizer.offerFeatureWhatever
threw exception while handling message
method = Optimizer/offerFeatureWhatever
message = { "name": "bob" }
exception = IllegalStateException
    at com.empowerops.volition.Optimizer.kt:123
    at xyz.java: 234
    at grpc.threadpoolInvoker.java:234
```

Maybe we can make "implement effective and simple logging" a volition issue?

---

>> * if `offer` returns false then we silently simply fail to send a reponse
>> * `offer` applies no back-pressure: if the `update` channel is clogged up then you could easily have this system drop messages when the load gets high.
> * if channel is null, at this moment it means it was unregistered.
> * unregistered != unavailable, so this will only cover the case where we somehow unregister the node during an evaluation. If the program dead or freezed, this will still try to do the offer.

I think you misunderstood me.

My concern is not about what happens when the channel is `null`. My concern if for what happens _when the channel is **full**_

do you understand what this means? if the channel is full then `offer` will return false. In this case **the message simply silently goes missing**, which seems like a really really bad behaviour.

The concept of back pressure is similar. The traditional solution to back-pressure is to avoid sending a response until you can be reasonably certain that either your job is done or will be done. So, in this case, when we send back a `Confirmation` message is that back pressure. 

What I think we should do is call `send` instead of offer. `send` on a Rendezvous channel will suspend the caller until somebody else calls `receive` and gets the message. This is a really nice behaviour because it means that the client wont see a confirmation message until we know that somebody is processing our message. 

Can you change this to `send` instead of `offer` please?

Reply:
- C/P doesn't have a change
- I wasn't trying to answer all your question here since I was also discovering new problem related to the similar fields when I was doing testing with oasis implementation, just couple thought and note on that
- send part which later talked about, I was just wondering when that could happen but I didn't follow up on that after we had a talk around this but no related to this particular topic. Anyway, it is changed now.